### DISCUSSION TITLE
Alternative way to use optimistic concurrency locking

### DISCUSSION THREAD

**Discussion by Tsimopak** [prisma/prisma#24993] [Q&A]
### Question

In the following page there is a suggestion how to solve concurrency issue by adding a `version` field and ask about it in the `where` clause when updating a model: 

```
const userEmail = 'alice@prisma.io'
const movieName = 'Hidden Figures'

// Find the first available seat
// availableSeat.version might be 0
const availableSeat = await client.seat.findFirst({
  where: {
    Movie: {
      name: movieName,
    },
    claimedBy: null,
  },
})

if (!availableSeat) {
  throw new Error(`Oh no! ${movieName} is all booked.`)
}

// Only mark the seat as claimed if the availableSeat.version
// matches the version we're updating. Additionally, increment the
// version when we perform this update so all other clients trying
// to book this same seat will have an outdated version.
const seats = await client.seat.updateMany({
  data: {
    claimedBy: userEmail,
    version: {
      increment: 1,
    },
  },
  where: {
    id: availableSeat.id,
    version: availableSeat.version, // This version field is the key; only claim seat if in-memory version matches database version, indicating that the field has not been updated
  },
})

if (seats.count === 0) {
  throw new Error(`That seat is already booked! Please try again.`)
}
```
Reference: https://www.prisma.io/docs/orm/prisma-client/queries/transactions#optimistic-concurrency-control

what I want to ask, is if instead of doing:
```
  where: {
    id: availableSeat.id,
    version: availableSeat.version, // This version field is the key; only claim seat if in-memory version matches database version, indicating that the field has not been updated
  },
```

I could have asked if claimedBy is still null, will it do the same trick?
so the solution would be like so:
```
  where: {
    id: availableSeat.id,
    claimedBy: null
  },
```

The only disadvantage compared to version is that if there were many swaps between claimedby and null, i would have not known about it, but that's fine by me I guess?

Note: @ludralph I appreciate you, but last time u gave me ChatGPT answer, so please, don't.

**Comment by matthewmueller** [comment]
Hey @Tsimopak,

That should be fine. There's some good videos on youtube about OCC, but you basically want to make sure that if two requests come in at the same time, the result will be the same no matter the order. 

This can get tricky if you have multiple queries because they could interleave. See: https://www.prisma.io/docs/orm/prisma-client/queries/transactions#scenario-reserving-a-seat-at-the-cinema for a good explanation of what can go wrong.

If you're unsure, you may also want to consider using an interactive transaction, which is less performant, but easier to conceptualize.

**Reply by Tsimopak** [reply]
Thanks matt for the answer :D Yeah I went through this example, so by that should be fine, you mean that it will be atomically handled in Prisma?

In other words, instead of having a version attribute, if i check in the where clause again the condition i was by the time I made the read, I do solve concurrency issues the same way I could have done with version?

**Reply by ludralph** [reply]
Hi @Tsimopak 

You may also want to take a look at this [article](https://dev.to/zenstack/how-to-build-a-high-concurrency-ticket-booking-system-with-prisma-184n) that shows the difference between using the `version` attribute and `claimedBy` in code and simulating same.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
